### PR TITLE
RM-9079 Add ASP.NET Core extension methods for static resource caching

### PR DESCRIPTION
**Pull request by a-ctor**
https://re-motion.atlassian.net/browse/RM-9079

### CODE REVIEW

File: Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    if (!remainingRequestPath.StartsWith(c_cacheKeyUrlIdentifier))
+      return false;
+
+    var indexOfSlash = remainingRequestPath.IndexOf('/', c_cacheKeyUrlIdentifier.Length);
+
+    // This should be impossible as we appended a slash earlier -> treat as no match to be defensive
+    if (indexOfSlash == -1)
+      return false;
+
+    var restPath = requestPathEndsWithSlash
```
Comment (MichaelKetting) [inline Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs:109]: "rest" -> "remaining"

File: Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    var requestPath = context.Request.Path;
+    newRequestPath = default;
+    cacheKey = default;
+
+    var requestPathEndsWithSlash = requestPath.HasValue && !requestPath.Value.EndsWith('/');
+    var requestPathWithTrailingSlash = requestPathEndsWithSlash
+        ? requestPath
+        : requestPath + new PathString("/");
+
+    if (!requestPathWithTrailingSlash.StartsWithSegments(_requestPath, out var rest) || !rest.HasValue)
```
Comment (MichaelKetting) [inline Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs:96]: rest -> remaining

File: Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+public class RemoveCacheKeyMiddleware
+{
+  private const string c_cacheKeyUrlIdentifier = "/cache_";
+  private const string c_removedCacheKeyItemsKey = "RemotionRemovedCacheKey";
+
+  /// <summary>
+  /// Tries to retrieve a cache key from <paramref name="context"/>, which was previously removed using <see cref="RemoveCacheKeyMiddleware"/>.
+  /// The cache key matches the value returned by <see cref="IStaticResourceCacheKeyProvider"/>.
+  /// </summary>
+  public static bool TryGetRemovedCacheKey (HttpContext context, [NotNullWhen(true)] out string? cacheKey)
```
Comment (MichaelKetting) [inline Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs:44]: do we ever need the actual value or would this be clearer as a simple boolean operation?

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

File: Remotion/Web/Core/Resources/CachingApplicationBuilderExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  /// <summary>
+  /// Enables static file serving with caching support for the current request path.
+  /// Replaces the call to <see cref="StaticFileExtensions.UseStaticFiles(Microsoft.AspNetCore.Builder.IApplicationBuilder)"/>.
+  /// </summary>
+  /// <remarks>
+  /// Requires <see cref="ResourcePathBuilderBasedStaticResourceCacheKeyProvider"/> to be useful.
+  /// </remarks>
+  public static IApplicationBuilder UseRemotionStaticFiles (this IApplicationBuilder builder, StaticFileOptions options, int cacheMaxAge = 604_800)
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Web/Core/Resources/CachingApplicationBuilderExtensions.cs:51]: Looking at this again I chose to introduce own own settings object `StaticFileCachingOptions` with init-only properties. I also converted the max age to a `TimeSpan`, and move the `Cache-Control` header generation over there. This makes the options more explicit and clear, while allowing the customization of the header value via subclassing.

File: Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+/// <summary>
+/// ASP.NET Core middleware that removes the cache key generated by <see cref="ResourcePathBuilderBasedStaticResourceCacheKeyProvider"/> from the URL.
+/// The URL is rewritten and the cache key is made available to any further middleware.
+/// </summary>
+/// <seealso cref="CachingApplicationBuilderExtensions.UseRemotionStaticFiles"/>
+/// <seealso cref="CachingApplicationBuilderExtensions.UseRemoveCacheKeyMiddleware"/>
+/// <remarks>
+/// The preferred way of using this middleware is <see cref="CachingApplicationBuilderExtensions"/>.<see cref="CachingApplicationBuilderExtensions.UseRemotionStaticFiles"/>.
+/// </remarks>
+public class RemoveCacheKeyMiddleware
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs:35]: It is, otherwise it would not compile :D

File: Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs (hunk 3 above)
Comment (a-ctor) [reply Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs:44]: We don't need it but I think it is a good to leave options open so I would keep it here.

File: Remotion/Web/Core/Resources/CachingApplicationBuilderExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  /// <summary>
+  /// Removes the cache key generated by <see cref="ResourcePathBuilderBasedStaticResourceCacheKeyProvider"/>.
+  /// Prefer using <see cref="UseRemotionStaticFiles"/> instead, as this methods is intended for advanced scenarios.
+  /// </summary>
+  /// <remarks>
+  /// Requires <see cref="ResourcePathBuilderBasedStaticResourceCacheKeyProvider"/> to be useful.
+  /// Only removes the cache key from the URL but does not change the request in any other way.
+  /// Other middleware logic can check if a cache key was removed using <see cref="RemoveCacheKeyMiddleware"/>.<see cref="RemoveCacheKeyMiddleware.TryGetRemovedCacheKey"/>.
+  /// </remarks>
+  public static IApplicationBuilder UseRemoveCacheKeyMiddleware (this IApplicationBuilder builder, PathString requestPath)
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Web/Core/Resources/CachingApplicationBuilderExtensions.cs:37]: Yeah probably does not have a good reason to exist anymore as most of the customization is possible in the wrapped `OnPrepareResponse` anyways.

**Comment by a-ctor** [comment]
@MichaelKetting I have addressed all of your review comments. I also discovered that the `UseRemotionStaticFilesCacheKeyRemovalMiddleware` was better of being a rewrite rule than a middleware. In terms of code the changes are minimal but it makes it easier to reuse if needed in the future. Another benefit is that we now use the official rewrite middleware, which handles edge cases like routing for us.

File: Remotion/Web/Core/Resources/CachingApplicationBuilderExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  /// </summary>
+  /// <remarks>
+  /// Requires <see cref="ResourcePathBuilderBasedStaticResourceCacheKeyProvider"/> to be useful.
+  /// </remarks>
+  public static IApplicationBuilder UseRemotionStaticFiles (this IApplicationBuilder builder, StaticFileOptions options, StaticFileCachingOptions cachingOptions)
+  {
+    ArgumentUtility.CheckNotNull(nameof(builder), builder);
+    ArgumentUtility.CheckNotNull(nameof(options), options);
+
+    builder.UseRewriter(new RewriteOptions().Add(new RemotionStaticFilesCacheKeyRemovalRewriteRule(options.RequestPath, true)));
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Web/Core/Resources/CachingApplicationBuilderExtensions.cs:41]: It is possible to use the rewrite middleware multiple times. Web Tests are currently not possible because we don't run the test site with .NET, which is required for the middleware stuff to run.

File: Remotion/Web/UnitTests/Core/Resources/RemotionStaticFilesCacheKeyRemovalRewriteRuleTest.cs [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    Assert.That(httpContext.Request.Path, Is.EqualTo((PathString)path));
+    Assert.That(httpContext.Items.Count, Is.EqualTo(0));
+
+    var hasCacheKey = RemotionStaticFilesCacheKeyRemovalRewriteRule.TryGetRemovedCacheKey(httpContext, out _);
+    Assert.That(hasCacheKey, Is.False);
+  }
+
+  [Test]
+  [TestCase("/static/cache_/", "/static/", "")]
+  [TestCase("/static/cache_abc/", "/static/", "abc")]
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Web/UnitTests/Core/Resources/RemotionStaticFilesCacheKeyRemovalRewriteRuleTest.cs:72]: The rewrite rule tested in this file does not do anything cache related on its own. The static file middleware will only handle the request (and add caching) if the file exists. Otherwise, further middlewares in the chain are called. As such, `/static/` would not be handled by the static file middleware as it is not a file and will fail with 404 without adding any caching header.

File: Remotion/Web/Core/Resources/CachingApplicationBuilderExtensions.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply Remotion/Web/Core/Resources/CachingApplicationBuilderExtensions.cs]: > Web Tests are currently not possible because we don't run the test site with .NET, which is required for the middleware stuff to run.

Hence please create a ticket for re-motion 7 :)

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply Remotion/Web/Core/Resources/CachingApplicationBuilderExtensions.cs]: https://re-motion.atlassian.net/browse/RM-9132

File: Remotion/Web/UnitTests/Core/Resources/RemotionStaticFilesCacheKeyRemovalRewriteRuleTest.cs [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+{
+  [Test]
+  public void Constructor_RequestPathWithNoTrailingSlash_Throws ()
+  {
+    Assert.That(
+        () => new RemotionStaticFilesCacheKeyRemovalRewriteRule("/abc/", false),
+        Throws.ArgumentException.With.ArgumentExceptionMessageEqualTo(
+            "Request path must not end in a slash.",
+            "requestPath"));
+  }
```
Comment (MichaelKetting) [inline Remotion/Web/UnitTests/Core/Resources/RemotionStaticFilesCacheKeyRemovalRewriteRuleTest.cs:37]: You want to also add a test for the empty case?

**Review by MichaelKetting — APPROVED** [review]
